CMakeLists.txt

```diff
@@ -114,6 +117,41 @@ if(CODECOVERAGE)
     message( "-- Building with llvm Code Coverage Tools")
 endif()
 
+if(CPPCHECK)
+    find_program(cppcheck cppcheck HINTS "/usr/local/bin/cppcheck" REQUIRED)
+    message(STATUS "cppcheck ${cppcheck}")
+    if(cppcheck MATCHES "NOTFOUND")
+      message(FATAL_ERROR "failed to find cppcheck executable for CPPCHECK option")
+    endif()
+    # Create <cppcheck> work folder for whole program analysis, for faster analysis and to store some useful debug information
+    # Add cppcheck work folder and reports folder for cppcheck output.
+    file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/cppcheck/reports/)
+    # max number of threads = number of CPUs
+    include(ProcessorCount)
+    ProcessorCount(CPU_CORES)
+    # Report output is sent to stderr. To have the output in a single report file, run run-cppcheck.sh only to analyse results, for example:
+    # concord-bft/build > make bcstatetransfer_tests 2> cppcheck/reports/full_report.txt
+    # concord-bft/build > scripts/run-cppcheck.sh -p
+    set(CMAKE_CXX_CPPCHECK
+          "${cppcheck}"
+          "--enable=all"
+          "--inconclusive"
+          "--inline-suppr"
+          "--quiet"
+          "--std=c++17"
+          "--platform=unix64"
+          "--template=cppcheck1"
+          "--max-configs=1"
+          "--library=boost.cfg"
+          "--library=openssl.cfg"
+          "--library=googletest"
+          "--addon=threadsafety.py"
+          "--cppcheck-build-dir=${PROJECT_BINARY_DIR}/cppcheck/"
+          "--suppressions-list=${CMAKE_CURRENT_SOURCE_DIR}/.cppcheck/suppressions.txt"
+          "--exitcode-suppressions=${CMAKE_CURRENT_SOURCE_DIR}/.cppcheck/exitcode-suppressions.txt"
```

Comment:
we need to add `--error-exitcode=1` to enable compilation failure on cppcheck errors

Reply:
Assuming that there is 1 error in every file in a target, and there are N files, this requires 
3........N builds + fixes between.
As of now I think it will not be the best way , since we dont want to enable CI yet, so you cant get the big picture on all issues by looking at all report. I think for now we output all to full_report.txt (see confluence) and decide status there. See confluence page.